woocommerce/woocommerce-gateway-stripe
# Fix manual renewal subscriptions that use SEPA when disabling the Legacy experience

Fixes #3133

## Changes proposed in this Pull Request:

- Introduce a migrator class extending [WCS_Background_Repairer](https://github.com/Automattic/woocommerce-subscriptions-core/blob/trunk/includes/abstracts/abstract-wcs-background-repairer.php)
- Select the subscriptions using legacy SEPA tokens (those with the `stripe_sepa` gateway ID) to be migrated
- For each subscription, create an updated SEPA gateway token if none exists. 
    This isn't needed to fix the problem, but doing it for consistency for the shoppers. Assuming it could be strange for shoppers to see an active subscription but no tokens.
- Update the subscription payment method type to use the updated SEPA `stripe_sepa_debit` gateway ID. 
    This is what fixes the actual problem

## Testing instructions

ℹ️ Since 8.2.0, legacy SEPA payment methods aren't attached to the customer for some reason. Check out 8.1.0 to add legacy SEPA payment methods if you want to test processing renewals. I'll create an issue to check this out.

### Setup

- Install and activate [WC Subscriptions](https://woocommerce.com/products/woocommerce-subscriptions/)
- Create a subscription product
- Create a customer user to be used as the shopper
- Set the store currency to Euro 
- Create a page using the classic checkout if you don't have one already
- Set up webhooks
- Under the payment methods tab, enable SEPA. At `wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods`

### Create subscriptions with the Legacy SEPA gateway

1. Log in as the shopper
1. Go to My Account > Payment methods > Add payment method, at `/my-account/add-payment-method/`
1. Add three SEPA payment methods. 
    [Find test IBAN numbers here](https://docs.stripe.com/sources/sepa-debit#testing). `AT611904300234573201`, `DE89370400440532013000`, `EE382200221020145685`
1. Add a card payment method 
1. As the shopper, purchase the subscription product four times - one with each payment method.
    You'll need to use the shortcode checkout. Legacy SEPA doesn't show up in Blocks
1. Confirm that
    - All subscriptions are active
    - Each of them display the selected payment method
    <img width="700" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/3aa7e2b2-1627-462f-81cc-b35d87eaaaf8">
1. Go to the Edit page for each subscription at `/wp-admin/admin.php?page=wc-orders--shop_subscription&action=edit&id=< sub ID >`
1. Under "Payment method", confirm SEPA Direct Debit is displayed
    <img width="700" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/6306ea89-d51c-40ef-8cc5-e6133c0d0264">

At this point:
- Going into the database > the wc_orders table > these SEPA subscriptions will have `stripe_sepa` under the `payment_method` column
- The Legacy SEPA payment gateway is enabled on the Stripe extension, which uses the `stripe_sepa` gateway ID 

### See the issue

1. Disable the Legacy experience
1. As a merchant, go to the Edit page for each subscription
1. Notice that under "Payment method", it now says "Manual renewal"

At this point:
- The Legacy SEPA payment gateway (`stripe_sepa`) is disabled
- The Updated SEPA payment gateway (`stripe_sepa_debit`) is enabled instead
- WooCommerce flags the payment gateway used for the subscription is disabled, thus Subscription changes the subscription to Manual renewal

### Confirm the right subscriptions are queued to be updated - actually testing the fix now

1. As a merchant, go to the Scheduled Action page
1. Confirm that there's a scheduled action called `wc_stripe_schedule_subscriptions_legacy_sepa_token_repairs`
    <img width="700" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/b308f7fe-2356-4213-8ab9-06a765c2b139"> 
1. Click on "Run now", or wait for the action to run
    This action will schedule a new action for each subscription that needs to be updated
1. If you let it execute on its own, confirm that another `wc_stripe_schedule_subscriptions_legacy_sepa_token_repairs` doesn't get scheduled for later
1. Search for actions called `wc_stripe_subscriptions_legacy_sepa_token_repair`
1. Confirm there are three actions scheduled. 
1. Confirm the ID of the subscription each one of them targets belongs to a subscription purchased with the Legacy SEPA gateway
1. Confirm that the subscription purchased with the card isn't listed
    <img width="700" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/932c291e-9c4a-48f4-855a-f3a29c796496">


### Confirm the subscriptions' payment method gets updated

1. As a merchant, go to the Scheduled Action page
1. Search for actions called `wc_stripe_subscriptions_legacy_sepa_token_repair`
1. Run the actions for each subscription. Or, preferably, wait for them to run
1. As a merchant, go to the Edit page for each subscription
1. Confirm that under "Payment method", it now says "SEPA Direct Debit"
1. On the WooCommerce logs, at `wp-admin/admin.php?page=wc-status&tab=logs`, confirm that there are entries for each subscription that was updated

What we're doing for each subscription:
- We retrieve all of the `stripe_sepa_debit` tokens for the user that owns the subscription. 
- When calling `WC_Payment_Tokens::get_customer_tokens()`, we create tokens using the Updated SEPA gateway (`stripe_sepa_debit`) from the Legacy tokens (`stripe_sepa`), [over here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.3.0/includes/class-wc-stripe-payment-tokens.php#L316)
- We update the `payment_method` of the subscription, from `stripe_sepa` to `stripe_sepa_debit`, which matches the gateway ID of the Updated SEPA gateway
- WooCommerce no longer flags the payment gateway used for the subscription as disabled; thus Subscription is no longer flagged under "Manual renewal"


### Unhappy paths

- No action must be scheduled when WC_Subscriptions isn't active
- No action must be scheduled when the Legacy experience is enabled

---


**Post merge**
